### ISSUE TITLE
2019-06-07-action2: test if `@prefix: false` without `@id` works as expected (Pierre-Antoine Champin)

### ISSUE THREAD

**Issue by iherman** [w3c/json-ld-wg#87]
test if `@prefix: false` without `@id` works as expected (Pierre-Antoine Champin) ([see details](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-06-07-json-ld#action2))

Cc: @pchampin

**Comment by pchampin** [comment]
TL;DR: it doesn't work, and making it work would involve heavy changes in the spec

## Current state of the spec

First: when `@prefix` is false, `@id` is mandatory (and would still be with the addition proposed in w3c/json-ld-api#76, for that matter).

Second, making `@id` optional in this situation would not make much sense, because **currently `@prefix` is only used for compaction, *not* for expansion**.

It means that
```json
{
  "@context": {
    "@version": 1.1,
    "foo": {
      "@id": "http://example.org/foo",
      "@prefix": false
    }
  },
  "foo:bar": "baz"
}
```
expands to 
```json
[
  {
    "http://example.org/foobar": [
      {
        "@value": "baz"
      }
    ]
  }
]
```
([try it in the playground](https://json-ld.org/playground/#startTab=tab-expanded&json-ld=%7B%22%40context%22%3A%7B%22%40version%22%3A1.1%2C%22foo%22%3A%7B%22%40id%22%3A%22http%3A%2F%2Fexample.org%2Ffoo%22%2C%22%40prefix%22%3Afalse%7D%7D%2C%22foo%3Abar%22%3A%22baz%22%7D))

The compaction algorithm find a term definition for `foo` and expands `foo:bar` as a compact IRI, regardless of the presence of `"@prefix": false` in its term definition.

In my opinion, there are two distinct things that need to be discussed.

## 1. Should we change the semantics of `@prefix`, so that compaction also takes it into account?

After this change, the example above would expand to
```json
[
  {
    "foo:bar": [
      {
        "@value": "baz"
      }
    ]
  }
]
```
(i.e. `foo:bar` would be considered as an absolute IRI)

On the one hand, I think this would be globally more consistent and easier to understand for users.

On the other hand, I don't know how deep the implications of that change would be (but I suspect that a can of worms lies there).

## 2. Do we want to solve w3c/json-ld-syntax#177 with `@prefix`?

If we answer "yes" to the previous question, then it could make sense to allow a term definition of the form `"mailto": { "@prefix": false }` or `"mailto": { "@prefix": false, "@id": null }`. This would prevent the use of the given term *either* as a prefix or as a vocabulary term.

However, as we have pointed out [during the meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-06-07-json-ld.html#section3-2), robustly "protecting" an IRI scheme that way is trickier than it looks, because of the complex interactions between the different kinds of scoped contexts and `@protected`. So users may use the `"@prefix": false` feature and think they are safe, while actually they are not.

Also, I found out another problem (w3c/json-ld-api#104) that arise from the ambiguity between compact IRIs and some absolute IRIs...

So I think we should rather tackle the problem at the root, and add a way to express an absolute IRI and make it clear that it is *not* a compact IRI. This could be done in the line of "Proposed solution 3" at w3c/json-ld-api#104.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-06-14-json-ld#section3-1).

- No actions or resolutions
<details><summary><i class="differentiate">View the transcript</i></summary>
<b>2019-06-07-action2:</b> test if <code>@prefix:</code> false without <code>@id</code> works as expected (Pierre-Antoine Champin) #87<br/>
Benjamin Young: issue #87 <a href="https://github.com/w3c/json-ld-wg/issues/87">https://github.com/w3c/json-ld-wg/issues/87</a><br/>
Benjamin Young: born from #76 <a href="https://github.com/w3c/json-ld-api/issues/76">https://github.com/w3c/json-ld-api/issues/76</a><br/>
Ivan Herman: original-original: <a href="https://github.com/w3c/json-ld-syntax/issues/177">https://github.com/w3c/json-ld-syntax/issues/177</a><br/>
Benjamin Young: next two issues are related<br/>
Pierre-Antoine Champin: some background: it was possible to define certain URI schemes as prefixes<br/>
… leading to confusing them with the start of CURIs<br/>
… I tried to alter <code>@prefix</code> to avoid this<br/>
… but that turns out not to be possible<br/>
… . so <code>@prefix</code> must be accompanied by <code>@id</code><br/>
… even if you accept ruben’s alteration, it is still not simple<br/>
… <code>@prefix</code> is only used during compaction as a hint to the algo<br/>
… but never it is used in expansion<br/>
… to block interpretation as a CURI<br/>
… so doing this with <code>@prefix</code> would radically change the meaning of <code>@prefix,</code> to add an effect during expansion<br/>
… this thing will come back and bite other people than us<br/>
Adam Soroka: .. users would expect <code>@prefix</code> to work both ways<br/>
Pierre-Antoine Champin: but that may have real effects on implementors<br/>
Gregg Kellogg: yes, <code>@prefixis</code> only meaningful during compaction<br/>
… in 1.0 any prefix could be used<br/>
… so we made some changes<br/>
… including restrictions and a keyword <code>@prefix</code> to overcome them<br/>
… we do run into 1.0 compatibility issues here<br/>
… but this is something we just didn’t complete<br/>
… it’s not a major rewrite<br/>
… now we are forcing the use of expanded term defns for things that used to be simple prefixes<br/>
Dave Longley: +1 to what gregg is saying … it seems to mostly be restricted to very local change in IRI expansion where we need one extra flag check<br/>
Gregg Kellogg: the impact isn’t too bad on the algos<br/>
… we did make changes to limit the use of prefixes during compaction, we should make the same changes for expansion<br/>
Ivan Herman: I thikn this issue arose in a different ticket ^^^^<br/>
… I have said before that we shouldn’t add too many things to the spec<br/>
… do we really have to do anything about #177?<br/>
… does it occur significantly in 1.0 usage?<br/>
… do we need to spend two meetings on it?<br/>
… i don’t think so. my feeling is that we should either close or defer<br/>
… we struggled whether this is really a security issue<br/>
… it can be awkward<br/>
… in several years it did not bite us<br/>
Benjamin Young: we haven’t had <code>@prefix</code> very long either!<br/>
Ivan Herman: but it’s not a matter of just <code>@prefix</code><br/>
Pierre-Antoine Champin: I agree with ivan<br/>
… I’m not convinced that this is a security issue<br/>
… . I’m not convinced that <code>@prefix</code> would be a good way to solve it if it were<br/>
Adam Soroka: .. getting this right with <code>@prefix</code> would be challenging<br/>
Pierre-Antoine Champin: <code>@prefix</code> was introduced for other reasons<br/>
… the way we have it now is counterintuitive<br/>
… we were all talking last week about using it this way without even realizing we couldn’t. that’s not good!<br/>
Gregg Kellogg: when we introduced the restrictions for prefixes, we overlooked expansion<br/>
… we left the job partly done<br/>
Ivan Herman: we are adding features and making the spec more complicated<br/>
Gregg Kellogg: is inconsistency complicated?<br/>
Ivan Herman: we are complicating the spec a few months before we try to get to CR<br/>
Dave Longley: I sympathize with ivan<br/>
… if we consider this a security issue, then <code>@prefix</code> is not how we should address it<br/>
… we could instead introduce a rule that you don’t expand prefixes when <code>@vocab</code> = true<br/>
Gregg Kellogg: I agree with dlongley; that would solve the problem but burn all the fields<br/>
Ivan Herman: I propose that we defer one of the two<br/>
Gregg Kellogg: I don’t think we can roll back prefix<br/>
… it fixes actual errors in 1.0<br/>
Pierre-Antoine Champin: I suggest we make no decision before the next agenda item<br/>
Gregg Kellogg: I can do a speculative PR<br/>
Benjamin Young: let’s go to the next issue, discuss, then come back to this<br/>
</details>
